core/src/backend/audio.rs

```diff
@@ -137,18 +142,247 @@ impl AudioBackend for NullAudioBackend {
+        if self.sounds.len() < Self::MAX_SOUNDS {
+            let handle = audio.start_sound(sound, settings).ok()?;
+            let instance = SoundInstance {
+                sound,
+                instance: handle,
+                display_object,
+                avm1_object,
+            };
+            audio.set_sound_transform(handle, self.transform_for_sound(&instance));
+            self.sounds.push(instance);
+            Some(handle)
+        } else {
+            None
+        }
+    }
+
+    pub fn stop_sound(&mut self, audio: &mut dyn AudioBackend, instance: SoundInstanceHandle) {
+        if let Some(i) = self
+            .sounds
+            .iter()
+            .position(|other| other.instance == instance)
+        {
+            let instance = &self.sounds[i];
+            audio.stop_sound(instance.instance);
+            self.sounds.swap_remove(i);
+        }
+    }
+
+    pub fn stop_sounds_with_handle(&mut self, audio: &mut dyn AudioBackend, sound: SoundHandle) {
+        self.sounds.retain(move |other| {
+            if other.sound == sound {
+                audio.stop_sound(other.instance);
+                false
+            } else {
+                true
+            }
+        });
+    }
+
+    pub fn stop_sounds_with_display_object(
+        &mut self,
+        audio: &mut dyn AudioBackend,
+        display_object: DisplayObject<'gc>,
+    ) {
+        self.sounds.retain(move |sound| {
+            if let Some(other) = sound.display_object {
+                if DisplayObject::ptr_eq(other, display_object) {
+                    audio.stop_sound(sound.instance);
+                    return false;
+                }
+            }
+            true
+        });
+    }
+
+    pub fn stop_all_sounds(&mut self, audio: &mut dyn AudioBackend) {
+        self.sounds.clear();
+        audio.stop_all_sounds();
+    }
+
+    pub fn is_sound_playing_with_handle(&mut self, sound: SoundHandle) -> bool {
+        self.sounds.iter().any(|other| other.sound == sound)
+    }
+
+    pub fn start_stream(
+        &mut self,
+        audio: &mut dyn AudioBackend,
+        clip_id: swf::CharacterId,
+
+        frame: u16,
+        data: crate::tag_utils::SwfSlice,
+        stream_info: &swf::SoundStreamHead,
+    ) -> Option<AudioStreamHandle> {
+        audio.start_stream(clip_id, frame, data, stream_info).ok()
+    }
+
+    pub fn stop_stream(&mut self, audio: &mut dyn AudioBackend, handle: AudioStreamHandle) {
+        audio.stop_stream(handle)
+    }
+
+    pub fn global_sound_transform(&self) -> DisplayObjectSoundTransform {
+        self.global_sound_transform
+    }
+
+    pub fn set_global_sound_transform(
+        &mut self,
+        audio: &mut dyn AudioBackend,
+        sound_transform: DisplayObjectSoundTransform,
+    ) {
+        self.global_sound_transform = sound_transform;
+        self.update_sound_transforms(audio);
+    }
+
+    fn transform_for_sound(&self, sound: &SoundInstance<'gc>) -> SoundTransform {
```

Comment:
Difficult to say because there's no way to print out the concatenated sound transforms in Flash; but in general, Flash prefers integer operations when the data types are integers/fixed point (see #193). Could possibly verify this using some sequence of transforms that would get truncated to 0.

I can verify that the values of the actual SoundTransform data type are integers using the following, at least:
```
var s = new Sound();
s.setVolume(2147483648.5)
trace(s.getVolume()); // -2147483648
```
but this doesn't mean that the intermediate calculations aren't float.

Reply:
Ok, I verified that the calculations should be int as well:
```
var s1 = new Sound();
s1.setVolume(2);

var s2 = new Sound(clip);
s2.attachSound("snd");
s2.setVolume(50); // Audible if this is 50 (final vol = 1), inaudible if 49 (final vol truncates to 0)
s2.start();
```